### DISCUSSION TITLE
hooking up citar and pdf-tools?

### DISCUSSION THREAD

**Discussion by bdarcus** [emacs-citar/citar#476] [Ideas]
Adding this in case others are interested; an idea from a private email.

Is there some way to elegantly:

- search within some subset of (pdf) files (by, for example, author)
- browse through the results, perhaps narrow them, and then to open the selected result's pdf at its respective page number

It seems to me a solution that hooks up citar and pdf-tools would be the way to go, and specifically to exploit [pdf-occur](https://github.com/politza/pdf-tools/blob/c510442ab89c8a9e9881230eeb364f4663f59e76/lisp/pdf-occur.el).

![image](https://user-images.githubusercontent.com/1134/145401178-0e3700a9-b060-47b0-94aa-68994c30bdcd.png)

Edit: solution is currently on the wiki.

https://github.com/bdarcus/citar/wiki/Example-functions#search-contents-of-pdfs

**Comment by bdarcus** [comment]
@roshanshariff @localauthor - this is a pretty cool idea. Do either you work much with dired programmatically?

I think we just need a way to mark possible files to search based on the citar selections, and then run `pdf-occur-dired-do-search`.

PS - I added #477 for this, with the idea to put it on the wiki if we can figure out a simple and elegant solution.

**Reply by danielnemenyi** [reply]
I'm trying [this out](https://github.com/bdarcus/citar/wiki/Example-functions#search-contents-of-pdfs) but I'm not seeing the relevant option to search pdfs listed after calling `citar-act` or `citar-act-all`. Any thoughts on what I might be missing?

![image](https://github.com/emacs-citar/citar/assets/8701448/a551f8a5-061f-48d6-8757-c4125fcae5ab)

**Reply by bdarcus** [reply]
You would have to add the command to your embark map for it to be recognized.

Does it work from `M-x`?

**Comment by danielnemenyi** [comment]
> You would have to add the command to your embark map for it to be recognized.

Aha, makes sense

> Does it work from M-x?

Ok, yes, it launches but errors on the function `citar-file--files-for-multiple-entries`, which seems to have been removed in ce51325?

**Comment by bdarcus** [comment]
Yeah, a lot has changed since I put that on the wiki.

My hunch is a revised example would be simpler.

If you manage to sort it out, can you please update the wiki?

Otherwise, I'll take a look when I get a chance.

**Reply by danielnemenyi** [reply]
Sure, I'll update the wiki if I work something out. Thanks

**Reply by bdarcus** [reply]
Take a look at `citar-get-files`.

It returns a hash-table of all files by key. You just need to filter for the PDFs, which I forget how to do ATM :-)

It might be you first just let-bind the list of all files for the selected citekeys, and then use `seq-filter` on that?

Something like this:

```elisp
(defun ex/search-pdf-contents ()
  "Search pdfs."
  (interactive)
  (let* ((selected (citar-select-refs))
        (files (citar-get-files selected))
        (pdfs (seq-filter ...)) ;; filter the PDFs here
        (search-str (read-string "Search string: ")))
  (pdf-occur-search pdfs search-str t)))
```

@roshanshariff - is there a better way to get the list of PDFs for the selected references here?

I haven't been working with this codebase much for awhile, and most of my coding in that time has been in Rust!

**Reply by roshanshariff** [reply]
In general, yes, that is the way to go. If you're only looking at files that are stored in library directories (rather than coming from the `file` field inside the bibliography file, say) you can also do something like this:
```elisp
(let ((citar-library-file-extensions '("pdf")))
  (citar-get-files key-or-keys))
```
This will filter the files as they're being searched for in the library directories, instead of finding them all and then keeping only the PDFs post-hoc.
